kube_downscaler/scaler.py

```diff
@@ -123,6 +124,86 @@ def autoscale_resource(resource: pykube.objects.NamespacedAPIObject, upscale_per
+        original_nodeselector = daemonset.annotations.get(ORIGINAL_NODESELECTOR_ANNOTATION)
+
+        if exclude and not original_nodeselector:
+            logger.debug('%s %s/%s was excluded', daemonset.kind, daemonset.namespace, daemonset.name)
+        else:
+            nodeselector = None
+            if 'nodeSelector' in daemonset.obj['spec']['template']['spec']:
+                nodeselector = daemonset.obj.get('spec')['template']['spec']['nodeSelector']
+            ignore = False
+
+            upscale_period = daemonset.annotations.get(UPSCALE_PERIOD_ANNOTATION, upscale_period)
+            downscale_period = daemonset.annotations.get(
+                DOWNSCALE_PERIOD_ANNOTATION, downscale_period)
+            if forced_uptime or (exclude and original_nodeselector):
+                uptime = "forced"
+                downtime = "ignored"
+                is_uptime = True
+            elif upscale_period != 'never' or downscale_period != 'never':
+                uptime = upscale_period
+                downtime = downscale_period
+                if helper.matches_time_spec(now, uptime) and helper.matches_time_spec(now, downtime):
+                    logger.debug('Upscale and downscale periods overlap, do nothing')
+                    ignore = True
+                elif helper.matches_time_spec(now, uptime):
+                    is_uptime = True
+                elif helper.matches_time_spec(now, downtime):
+                    is_uptime = False
+                else:
+                    ignore = True
+                logger.debug('Periods checked: upscale=%s, downscale=%s, ignore=%s, is_uptime=%s',
+                             upscale_period, downscale_period, ignore, is_uptime)
+            else:
+                uptime = daemonset.annotations.get(UPTIME_ANNOTATION, default_uptime)
+                downtime = daemonset.annotations.get(DOWNTIME_ANNOTATION, default_downtime)
+                is_uptime = helper.matches_time_spec(
+                    now, uptime) and not helper.matches_time_spec(now, downtime)
+
+            logger.debug('%s %s/%s has "%s" nodeSelector (original: %s, uptime: %s)',
+                         daemonset.kind, daemonset.namespace, daemonset.name, nodeselector, original_nodeselector, uptime)
+            update_needed = False
+
+            if not ignore and is_uptime and nodeselector is not None \
+                    and 'non-existing' in daemonset.obj['spec']['template']['spec']['nodeSelector'] \
+                    and daemonset.obj['spec']['template']['spec']['nodeSelector']['non-existing'] == 'true':
+                logger.info('Scaling up %s %s/%s from nodeSelector "%s" to "%s" (uptime: %s, downtime: %s)',
+                            daemonset.kind, daemonset.namespace, daemonset.name, nodeselector, original_nodeselector,
+                            uptime, downtime)
+                daemonset.obj['spec']['template']['spec']['nodeSelector'] = original_nodeselector
```

Comment:
`original_nodeselector` is a string and `nodeSelector` should be a `dict` --- unsure how this would work

Reply:
Yes, I just tested and there are some issues if there's already a nodeSelector defined and looks like `original_nodeselector` is not working properly. I'll update the PR with the fix and full test coverage of this. Thanks

Reply:
New tests, and fix is still in progress (about 50% done) but it will take a couple of more weeks since I don't have the time :(